// xmem-ai/dist/src/client.js
"use strict";
/**
 * XMem TypeScript client.
 *
 * ```ts
 * import { XMemClient } from "@xmem/sdk";
 *
 * const client = new XMemClient("http://localhost:8000", "sk-...");
 * const result = await client.ingest({ user_query: "I love hiking", user_id: "u1" });
 * const answer = await client.retrieve({ query: "hobbies", user_id: "u1" });
 * console.log(answer.answer);
 * ```
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.XMemClient = void 0;
const network_1 = require("./network");
class XMemClient {
    constructor(apiUrl, apiKey) {
        this.baseUrl = (apiUrl || process.env.XMEM_API_URL || "http://localhost:8000").replace(/\/$/, "");
        const key = apiKey || process.env.XMEM_API_KEY || "";
        this.headers = {
            "Content-Type": "application/json",
            ...(key ? { Authorization: `Bearer ${key}` } : {}),
        };
    }
    async req(method, path, body) {
        return (0, network_1.request)(this.baseUrl, path, {
            method,
            headers: this.headers,
            body: body ? JSON.stringify(body) : undefined,
        });
    }
    // ── Health ───────────────────────────────────────────────────────
    async ping() {
        try {
            const env = await this.req("GET", "/health");
            return env.data;
        }
        catch {
            return { status: "unreachable", pipelines_ready: false, version: "" };
        }
    }
    async isReady() {
        return (await this.ping()).pipelines_ready;
    }
    // ── Ingest ───────────────────────────────────────────────────────
    async ingest(params) {
        const env = await this.req("POST", "/v1/memory/ingest", params);
        return { ...env.data, request_id: env.request_id, elapsed_ms: env.elapsed_ms };
    }
    // ── Retrieve ─────────────────────────────────────────────────────
    async retrieve(params) {
        const env = await this.req("POST", "/v1/memory/retrieve", params);
        return { ...env.data, request_id: env.request_id, elapsed_ms: env.elapsed_ms };
    }
    // ── Search ───────────────────────────────────────────────────────
    async search(params) {
        const env = await this.req("POST", "/v1/memory/search", params);
        return { ...env.data, request_id: env.request_id, elapsed_ms: env.elapsed_ms };
    }
}
exports.XMemClient = XMemClient;
