// xmem-ai/dist/src/error.js
"use strict";
/**
 * Error hierarchy for the XMem TypeScript SDK.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConnectionError = exports.NotReadyError = exports.ValidationError = exports.ServerError = exports.RateLimitError = exports.AuthenticationError = exports.XMemSDKError = void 0;
class XMemSDKError extends Error {
    constructor(message, statusCode, requestId) {
        super(message);
        this.name = "XMemSDKError";
        this.statusCode = statusCode;
        this.requestId = requestId;
    }
}
exports.XMemSDKError = XMemSDKError;
class AuthenticationError extends XMemSDKError {
    constructor(message, statusCode, requestId) {
        super(message, statusCode, requestId);
        this.name = "AuthenticationError";
    }
}
exports.AuthenticationError = AuthenticationError;
class RateLimitError extends XMemSDKError {
    constructor(message, statusCode, requestId, retryAfter) {
        super(message, statusCode, requestId);
        this.name = "RateLimitError";
        this.retryAfter = retryAfter;
    }
}
exports.RateLimitError = RateLimitError;
class ServerError extends XMemSDKError {
    constructor(message, statusCode, requestId) {
        super(message, statusCode, requestId);
        this.name = "ServerError";
    }
}
exports.ServerError = ServerError;
class ValidationError extends XMemSDKError {
    constructor(message, statusCode, requestId) {
        super(message, statusCode, requestId);
        this.name = "ValidationError";
    }
}
exports.ValidationError = ValidationError;
class NotReadyError extends XMemSDKError {
    constructor(message, statusCode, requestId) {
        super(message, statusCode, requestId);
        this.name = "NotReadyError";
    }
}
exports.NotReadyError = NotReadyError;
class ConnectionError extends XMemSDKError {
    constructor(message) {
        super(message);
        this.name = "ConnectionError";
    }
}
exports.ConnectionError = ConnectionError;
