// xmem-ai/dist/src/network.js
"use strict";
/**
 * Internal HTTP transport for the XMem TypeScript SDK.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.request = request;
const error_1 = require("./error");
function mapError(status, body) {
    const msg = body.error || body.detail || `HTTP ${status}`;
    const rid = body.request_id;
    if (status === 401 || status === 403)
        return new error_1.AuthenticationError(msg, status, rid);
    if (status === 422)
        return new error_1.ValidationError(msg, status, rid);
    if (status === 429)
        return new error_1.RateLimitError(msg, status, rid, Number(body.retry_after ?? 60));
    if (status === 503)
        return new error_1.NotReadyError(msg, status, rid);
    if (status >= 500)
        return new error_1.ServerError(msg, status, rid);
    return new error_1.XMemSDKError(msg, status, rid);
}
async function request(baseUrl, path, init) {
    let response;
    try {
        response = await fetch(`${baseUrl}${path}`, init);
    }
    catch (err) {
        throw new error_1.ConnectionError(`Connection failed: ${err}`);
    }
    let body;
    try {
        body = (await response.json());
    }
    catch {
        if (response.status >= 400) {
            throw mapError(response.status, { error: await response.text().catch(() => "") });
        }
        throw new error_1.XMemSDKError(`Unparseable response (HTTP ${response.status})`);
    }
    if (response.status >= 400) {
        throw mapError(response.status, body);
    }
    if (body.status === "error") {
        throw mapError(response.status || 500, body);
    }
    return {
        status: "ok",
        data: body.data,
        error: body.error,
        request_id: body.request_id,
        elapsed_ms: body.elapsed_ms,
    };
}
