// xmem-ai/src/client.ts
/**
 * XMem TypeScript client.
 *
 * ```ts
 * import { XMemClient } from "@xmem/sdk";
 *
 * const client = new XMemClient("http://localhost:8000", "sk-...");
 * const result = await client.ingest({ user_query: "I love hiking", user_id: "u1" });
 * const answer = await client.retrieve({ query: "hobbies", user_id: "u1" });
 * console.log(answer.answer);
 * ```
 */

import { request } from "./network";
import type {
  APIEnvelope,
  HealthStatus,
  IngestParams,
  IngestResult,
  RetrieveParams,
  RetrieveResult,
  SearchParams,
  SearchResult,
  SourceRecord,
} from "./types";

export class XMemClient {
  private readonly baseUrl: string;
  private readonly headers: HeadersInit;

  constructor(apiUrl?: string, apiKey?: string) {
    this.baseUrl = (apiUrl || process.env.XMEM_API_URL || "http://localhost:8000").replace(
      /\/$/,
      "",
    );

    const key = apiKey || process.env.XMEM_API_KEY || "";
    this.headers = {
      "Content-Type": "application/json",
      ...(key ? { Authorization: `Bearer ${key}` } : {}),
    };
  }

  private async req<T = Record<string, unknown>>(
    method: string,
    path: string,
    body?: unknown,
  ): Promise<APIEnvelope<T>> {
    return request<T>(this.baseUrl, path, {
      method,
      headers: this.headers,
      body: body ? JSON.stringify(body) : undefined,
    });
  }

  // ── Health ───────────────────────────────────────────────────────

  async ping(): Promise<HealthStatus> {
    try {
      const env = await this.req<HealthStatus>("GET", "/health");
      return env.data!;
    } catch {
      return { status: "unreachable", pipelines_ready: false, version: "" };
    }
  }

  async isReady(): Promise<boolean> {
    return (await this.ping()).pipelines_ready;
  }

  // ── Ingest ───────────────────────────────────────────────────────

  async ingest(params: IngestParams): Promise<IngestResult> {
    const env = await this.req<IngestResult>("POST", "/v1/memory/ingest", params);
    return { ...env.data!, request_id: env.request_id, elapsed_ms: env.elapsed_ms };
  }

  // ── Retrieve ─────────────────────────────────────────────────────

  async retrieve(params: RetrieveParams): Promise<RetrieveResult> {
    const env = await this.req<RetrieveResult>("POST", "/v1/memory/retrieve", params);
    return { ...env.data!, request_id: env.request_id, elapsed_ms: env.elapsed_ms };
  }

  // ── Search ───────────────────────────────────────────────────────

  async search(params: SearchParams): Promise<SearchResult> {
    const env = await this.req<SearchResult>("POST", "/v1/memory/search", params);
    return { ...env.data!, request_id: env.request_id, elapsed_ms: env.elapsed_ms };
  }
}
