// xmem-ai/src/error.ts
/**
 * Error hierarchy for the XMem TypeScript SDK.
 */

export class XMemSDKError extends Error {
  public readonly statusCode?: number;
  public readonly requestId?: string;

  constructor(message: string, statusCode?: number, requestId?: string) {
    super(message);
    this.name = "XMemSDKError";
    this.statusCode = statusCode;
    this.requestId = requestId;
  }
}

export class AuthenticationError extends XMemSDKError {
  constructor(message: string, statusCode?: number, requestId?: string) {
    super(message, statusCode, requestId);
    this.name = "AuthenticationError";
  }
}

export class RateLimitError extends XMemSDKError {
  public readonly retryAfter?: number;

  constructor(message: string, statusCode?: number, requestId?: string, retryAfter?: number) {
    super(message, statusCode, requestId);
    this.name = "RateLimitError";
    this.retryAfter = retryAfter;
  }
}

export class ServerError extends XMemSDKError {
  constructor(message: string, statusCode?: number, requestId?: string) {
    super(message, statusCode, requestId);
    this.name = "ServerError";
  }
}

export class ValidationError extends XMemSDKError {
  constructor(message: string, statusCode?: number, requestId?: string) {
    super(message, statusCode, requestId);
    this.name = "ValidationError";
  }
}

export class NotReadyError extends XMemSDKError {
  constructor(message: string, statusCode?: number, requestId?: string) {
    super(message, statusCode, requestId);
    this.name = "NotReadyError";
  }
}

export class ConnectionError extends XMemSDKError {
  constructor(message: string) {
    super(message);
    this.name = "ConnectionError";
  }
}
