// xmem-ai/src/network.ts
/**
 * Internal HTTP transport for the XMem TypeScript SDK.
 */

import type { APIEnvelope } from "./types";
import {
  AuthenticationError,
  ConnectionError,
  NotReadyError,
  RateLimitError,
  ServerError,
  ValidationError,
  XMemSDKError,
} from "./error";

function mapError(status: number, body: Record<string, unknown>): XMemSDKError {
  const msg = (body.error as string) || (body.detail as string) || `HTTP ${status}`;
  const rid = body.request_id as string | undefined;

  if (status === 401 || status === 403) return new AuthenticationError(msg, status, rid);
  if (status === 422) return new ValidationError(msg, status, rid);
  if (status === 429) return new RateLimitError(msg, status, rid, Number(body.retry_after ?? 60));
  if (status === 503) return new NotReadyError(msg, status, rid);
  if (status >= 500) return new ServerError(msg, status, rid);
  return new XMemSDKError(msg, status, rid);
}

export async function request<T = Record<string, unknown>>(
  baseUrl: string,
  path: string,
  init: RequestInit,
): Promise<APIEnvelope<T>> {
  let response: Response;
  try {
    response = await fetch(`${baseUrl}${path}`, init);
  } catch (err) {
    throw new ConnectionError(`Connection failed: ${err}`);
  }

  let body: Record<string, unknown>;
  try {
    body = (await response.json()) as Record<string, unknown>;
  } catch {
    if (response.status >= 400) {
      throw mapError(response.status, { error: await response.text().catch(() => "") });
    }
    throw new XMemSDKError(`Unparseable response (HTTP ${response.status})`);
  }

  if (response.status >= 400) {
    throw mapError(response.status, body);
  }

  if (body.status === "error") {
    throw mapError(response.status || 500, body);
  }

  // Handle both envelope-wrapped responses (with `data` field) and direct responses
  // Some endpoints like /health return data directly, others wrap it in { data: ... }
  const hasDataField = "data" in body && body.data !== undefined;
  const data = hasDataField ? (body.data as T) : (body as unknown as T);

  return {
    status: "ok",
    data,
    error: body.error as string | undefined,
    request_id: body.request_id as string | undefined,
    elapsed_ms: body.elapsed_ms as number | undefined,
  };
}
